### ISSUE TITLE
Git push

### ISSUE THREAD

**Issue by sagikazarmark** [dagger/dagger#7202]
### What are you trying to do?

First class or at least decent support for pushing to Git repositories.

### Why is this important to you?

Dagger already has first-class support for pulling from Git repositories (even if private repo support isn't trivial).

It's often important to push changes back to a repository. A couple use cases in mind:

- Regenerate SDKs from API descriptors upon change
- Update GitOps manifests when a new version is out

Unfortunately, feeding credentials into Git (whether it's SSH or HTTPS) is not trivial.

### How are you currently working around this?

There is already an existing example using [go-git](https://github.com/go-git/go-git): https://daggerverse.dev/mod/github.com/matipan/daggerverse/image-updater@7dafaf057c4524654903414fc3a7b5e2374f5695

Ideally, there should be a generic "git push" solution: whether it's built into Dagger or available as a separate module.

**Comment by shykes** [comment]
In case it's useful, I recently started a new git module. It doesn't support push, but has a clean and simple foundation to add it.

https://daggerverse.dev/mod/github.com/shykes/git

**Comment by sagikazarmark** [comment]
@shykes Sounds good!

What do you think would be the best way to implement authentication?

GitHub supports http+token auth, but SSH is probably still the most common way to authenticate with Git. CI systems (like GHA) probably work better with API tokens.

**Comment by shykes** [comment]
I would start simple, and add a `[]SSHKey` field to the `Repo` state, with assorted `WithSSHKey` and `WithSSHKeys` to add them. then a `Push()` function that makes sure to write the ssh keys to the right place in the container, and calls `git push`. Start with passing the argument as-is. Then, iterate from there.

**Comment by sagikazarmark** [comment]
Would it make sense to add `WithSSHKey` to the `Git` object instead (or as well) so they can be used to clone private repos as well?

**Comment by shykes** [comment]
> Would it make sense to add `WithSSHKey` to the `Git` object instead (or as well) so they can be used to clone private repos as well?

Yes that makes perfect sense.

**Comment by habib-source** [comment]
Any updates on this?

**Comment by jedevc** [comment]
@cwlbraa also mentioned he has a pretty strong use case for this in https://github.com/dagger/container-use :tada:

IMO, it feels like we need a handful of new APIs:
- Authoring new commits.
  
  Our git api should easily allow creating new commits, so as to avoid needing to do `GitRef.Tree`, mounting it, creating commits, and then calling `Directory.AsGit`. We probably want to have something like `GitRef.Append -> GitRef` that creates a new git-ref that points to a newly authored commit.

  I'm imagining something like
  ```graphql
  type GitRef {
    append(
      diff: Directory!

      author: string
      date: string
      # and other git commit parameteres
    ): GitRef!
  }
  ```

  I guess `diff` would be a `Directory` to apply on top of the previous state? The problem is this can't track *removal* of files. I wonder if we should natively support a "richer" `DirectoryDiff` type, something that can? I'm thinking maybe like a lazy "recipe" of file operations that could be all be joined together, and then applied to a directory?

- Pushing refs.

  This feels pretty simple - `GitRef.push` should allow pushing a ref. Something like:

  ```graphql
    type GitRef {
      push(
        # defaults to where they were cloned from
        remote: String,
        remoteRef: String,
        experimentalServiceHost: ServiceID
    
        # same as in Query.git
        sshKnownHosts: String = ""
        sshAuthSocket: SocketID
        httpAuthToken: SecretID
        httpAuthHeader: SecretID
      )
    }
    ```

    Only thing to note, the auth **must not be copied from the auth on the git object** - we need to pass all these in again (see the logic in https://github.com/dagger/dagger/issues/10572 - the security issue noted where a cloned repository could be pushed back to it's source needs to be avoided).

- Better access to host git repos.

  It would be nice to be able to turn *any* host git repo into a `GitRepo`/`GitRef`, using a new `Host.git`. Then, you'd potentially just be able to `Host.git(...).append(...).push()` to make changes and push directly back to the host.

   I think this would require another `GitBackend` implementation that maps directly to the host, which would connect to a git-daemon from the client via a attachable or something.

**Comment by cwlbraa** [comment]
@jedevc would it make sense to wrap the gitAuth information in a freestanding api object? that way, in code, you can easily share the auth information even though it needs to be provided in both spots?


```
gAuth = &GitAuthConfig{sshKnownHosts: x, sshAuthSocket: y}
Git(url: url, auth: gAuth).Branch(name: name).Append(xyz).Push(auth: gAuth)
```

... having typed it i now can't decide if that's more or less convenient.

**Comment by cwlbraa** [comment]
anyways, i'm fairly certain your proposal _almost_ fully covers container-use needs... it's possible i'm missing some details of how Tree() works, but [ideally we can allow folks to plop the tree directory into a container, do git cli operations on it with withexec, and then pull it back out as a GitRef and keep doing.](https://github.com/dagger/container-use/issues/45)

right now this doesn't exist in container-use, but we end up needing to prompt engineer agents to not try to install or use git inside containers.

**Comment by jedevc** [comment]
> would it make sense to wrap the gitAuth information in a freestanding api object?
> having typed it i now can't decide if that's more or less convenient.

Yeah 🤔 I also had this thought and couldn't decide either - great minds 😆 

We've got a few cases like this honestly, I also have wondered about whether we should have a username+password specialized secret type, which would be useful for things that interact with HTTP auth.

We probably wouldn't need to create them as a full-on object type, we could just have them be an input type, like [PortForward](https://github.com/dagger/dagger/blob/e063313d8ba8c3c14f43859ee43e161a2c7fe15d/core/schema/query.go#L60-L60). Since they're never returned by themselves.

> it's possible i'm missing some details of how Tree() works, but [ideally we can allow folks to plop the tree directory into a container, do git cli operations on it with withexec, and then pull it back out as a GitRef and keep doing.](https://github.com/dagger/container-use/issues/45)

Today you can do this:

```go
gitref := dag.Git("https://github.com/dagger/dagger.git").Head()
gitdir := gitref.Tree()

gitdir = dag.Container().From("git").
	WithWorkdir("/src").
	WithMountedDirectory("/src", gitdir).
	WithExec([]string{"git", /* do various git operations */}).
	Directory(".")
gitref = gitdir.AsGit().Head()
```

but it's just really verbose and annoying, and you need to spin up a container, bleh.

Being able to do all the "various git operations" (like `Append`) natively on `GitRef` types would hopefully avoid that.

**Comment by jedevc** [comment]
Just a note: realized I didn't explain why I thought it should be called `Append`. It's because `GitRef.commit` is already an API, which returns the SHA of the commit at that ref.

Other names are available, I don't think it *really* impacts the overall design though.
